### PR TITLE
New Scheduler- Productionisation Improvements

### PR DESCRIPTION
**Pull request by d80tb7**
Some improvements to the new Scheduler to aid in productionisation.

* Database schema now supports migrations
* Scheduler is now broken out into its own component
* Removed Job Errors and JobRunErrors from database
* Job Specs and Groups are now compressed

### CODE REVIEW

File: internal/armada/server.go [hunk 1]
```diff
@@ -438,15 +395,8 @@ func Serve(ctx context.Context, config *configuration.ArmadaConfig, healthChecks
...
-	// If the new Pulsar-driven scheduler is provided, run that.
-	// Otherwise run the legacy scheduler.
-	if newSchedulerApiServer != nil {
-		log.Info("Pulsar-based scheduler enabled")
-		api.RegisterAggregatedQueueServer(grpcServer, newSchedulerApiServer)
-	} else {
-		log.Info("legacy scheduler enabled")
-		api.RegisterAggregatedQueueServer(grpcServer, aggregatedQueueServer)
-	}
+	log.Info("legacy scheduler enabled")
```
[the remark this answers is not in the document]
Comment (d80tb7) [reply internal/armada/server.go:398]: this is done in #1828

File: internal/common/database/upsert.go [hunk 2]
```diff
@@ -36,7 +36,7 @@ import (
 //
 // )
 // I.e., it should omit everything before and after the "(" and ")", respectively.
-func Upsert(ctx context.Context, db *pgxpool.Pool, tableName string, schema string, records []interface{}) error {
+func Upsert[T any](ctx context.Context, db *pgxpool.Pool, tableName string, records []T) error {
```
[the remark this answers is not in the document]
Comment (d80tb7) [reply internal/common/database/upsert.go:39]: as discussed: `any` removes the reuirement to cast the array to interface

File: internal/scheduler/api.go [hunk 3]
```diff
@@ -86,33 +85,48 @@ func (srv *ExecutorApi) StreamingLeaseJobs(stream api.AggregatedQueue_StreamingL
...
-			sqlJobs[i].Groups,
-			sqlJobs[i].Queue,
-			sqlJobs[i].JobSet,
-			jobTime,
-			logJob,
-		)
-		if err != nil {
-			return err
+
+	srv.withDecompressor(func(decompressor compress.Decompressor) error {
```
[the remark this answers is not in the document]
Comment (d80tb7) [reply internal/scheduler/api.go:93]: we do- the `srv.withDecompressor` function handles the boilerplate of taking it out of the pool, and ensuring it is returned to the pool once we're done

File: cmd/scheduler/main.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	"github.com/G-Research/armada/internal/scheduler"
+	schedulerdb "github.com/G-Research/armada/internal/scheduler/database"
+)
+
+const (
+	CustomConfigLocation string = "config"
+	MigrateDatabase      string = "migrateDatabase"
+)
+
+func init() {
```
[the remark this answers is not in the document]
Comment (d80tb7) [reply cmd/scheduler/main.go:24]: yeah- I think that might be a better approach but I think I'd rather address how we do this run commands in a separate pr.  For now this is at least consistent with lookout/lookoutv2 which also migrate the db.

**Review by severinson — APPROVED** [review]

**Pull request merged**